### ISSUE TITLE
Support effort-driven scheduling

### ISSUE THREAD

**Issue by GoogleCodeExporter** [bardsoftware/ganttproject#83]
```
In effort-driven scheduling a length of a task depends on the amount of 
required work (in man-days or man-hours) and availability of resources. 

The basic rules is:
* If a task requires 3 man-days and a person who is doing this task is 50% 
available then the task must become 6 days long. 

Previous requests from SF issue tracker:
http://sourceforge.net/tracker/?
func=detail&aid=1102159&group_id=72728&atid=535430
http://sourceforge.net/tracker/?
func=detail&aid=1186184&group_id=72728&atid=535430
http://sourceforge.net/tracker/?
func=detail&aid=1250233&group_id=72728&atid=535430
```

Original issue reported on code.google.com by `dbarashev` on 9 Feb 2010 at 5:23

**Comment by JonSeagull** [comment]
I agree with the others...I was really loving this program until I hit the wall.  No resource leveling at all?  No effort driven task length?  For me this went immediately from "best program ever!" to worthless.

**Comment by joacimsvensson** [comment]
Agreed. No resource leveling kind and effort-driven schedulingof defeats the purpose of Ganttproject as a project MANAGEMENT tool.

I feels like something that should be a core functionality in any project management software. You have resources with allocated time, you have tasks - surely they should interact.

**Comment by CyrilleGuimezanes** [comment]
Hello!
Up, this is absolutly necesserary to make a real plannification...

**Comment by chuck990** [comment]
Just wanted to add another +1 to the request for resource-focused / effort-driven scheduling. It would be a major upgrade

**Comment by daniberc** [comment]
Hello,

I was looking for an alternative to MS Project, and I was very entusiastic with GanttProject, until I reach this issue.

From my point of view, to manage a project based in tasks,, I need the tasks to be affected by the number of ressources working on it, and their availabilty.

If a task as a duration of 10 days, and I assigned 2 ressources to it, I need that this task is reduced to 5 days.
And if I assigned only 1 ressources with a 50% dedication, the duration of the task should be of 20 days.

This features is a deal breaker, and it has not to be linked with the issue of the day-offs.

Hoping this issue would be take in account shortly.

**Comment by AlexDG89** [comment]
Hello,

+ 1 for me. It was useful until I read this. Everyone defines tasks as effort-based non day-based. This is a major issue.

**Comment by silveira** [comment]
+1

**Comment by vChavezB** [comment]
I am interested in contributing to this feature but it would be nice if I could get some of the following info from the devs/mantainers:

- Which source files are related to the options the user can modify via the Task property window?
- Where is the task duration calculated?  I saw something [here](https://github.com/bardsoftware/ganttproject/blob/e86cb8d475c7ebfcd143d07d0223c58bb3c034d1/ganttproject/src/main/java/net/sourceforge/ganttproject/chart/mouse/ChangeTaskProgressRuler.java#L54) but I am not sure.

**Comment by Exxerial** [comment]
+1 This issue has been ongoing since 2017 (5 years now). I think this is a major feature and should be priority task...

**Comment by Synochy** [comment]
+1 very interested by this feature

**Comment by TaniumOG** [comment]
+1

**Comment by primepeter** [comment]
@vChavezB did you made any progress on this?

**Comment by Natalie-the-technician** [comment]
I've implemented this in a fork, and it does what the original description asks for: task duration derived from effort and resource availability.

Available at https://github.com/Natalie-the-technician/ganttproject-effort-driven

**The rule from the issue body works as written.** Enter an effort in hours on a task, assign someone, and the duration follows from their hours per day, their utilisation percentage and the share of their time allocated to that task. Three man-days at 50% availability gives six days. Multiple resources add up; rounding is always upward with a one-day minimum.

Missing values are handled explicitly rather than defensively: no effort means the duration is left alone, so the feature is opt-in per task and existing projects are not disturbed.

**@silveira's point from 2022 is covered too.** There is a capacity levelling pass that spreads competing tasks until nobody is booked beyond their capacity, and it shows what it intends to move before moving it.

**@vChavezB — the entry points you asked about in 2022**, in case they are useful to anyone:

- `EffortDrivenDurationAlgorithm` — the calculation, registered in `AlgorithmCollection`
- `EffortDrivenTrigger` — fires on assignment changes, resource changes and resource removal; deliberately not on resource addition
- The properties live as ordinary custom columns, so projects stay readable in stock GanttProject
- The pass runs before the scheduler, so the scheduler can propagate the new dates

**Structure.** The work is split into branches so any part can be taken on its own: `effort-planning` is the scheduling work, `toggl-import` is optional time tracking on top, and `webdav-fixes` and `misc-fixes` contain only corrections to existing behaviour with no feature work at all. Each branch builds and passes its tests independently.

**Limits, plainly.** Used on real plans but not long-term, tested by one person. Building needs a JDK that ships JavaFX. There are known rough edges, listed in `FORK.md` — including one where stock GanttProject reports an error opening a project that uses a custom date column, which is a defect in the original and reported separately.

Offered as something that works today, not as a pull request. Happy to answer questions.

**Comment by dbarashev** [comment]
@Natalie-the-technician I really appreciate your contributions! Since your fork combines a few different updates, would you mind breaking it into focused, single-feature PRs? This  would help me to review and merge everything much faster—especially the effort-driven scheduling. Let us know if you have time for this, and we'd be happy to offer compensation for the effort required to split them up. Drop me an email if you're interested.

**Comment by Natalie-the-technician** [comment]
Thanks @dbarashev  — I'd be glad to.

The work is already split, which should make this straightforward. There are five branches, each building and passing its tests independently:

- `fork-base` — shared infrastructure plus three fixes to the original
- `effort-planning` — effort-driven scheduling, capacity levelling, recurring tasks, estimate quality
- `toggl-import` — Toggl integration, on top of `effort-planning`
- `webdav-fixes` — eleven WebDAV fixes, independent of the planning work
- `misc-fixes` — six UI fixes, independent as well

The separation is verified rather than assumed: introducing a deliberate cross-reference between sibling branches makes the build fail.

`effort-planning` is still larger than a single-feature PR, so I'd split it further — effort-driven duration first, since that is what this issue asks for, then levelling, recurrence and the estimate report separately. Tell me if you'd prefer a different cut.

**One thing you should know before I open anything.** While testing I hit a defect in the original that affects data, not just my fork:

Any custom property of class `DATE` or `DOUBLE` that has no value makes the task properties dialog fail on OK with a `NullPointerException` in `CustomColumnsValues.addCustomProperty` — `defStub.defaultValue!!` on a null default from `PropertyTypeEncoder`. Because `TaskProperties.save` runs the panels in sequence, the "General" tab is already written when it throws, and predecessors and resources are silently never saved. The user sees only a generic error.

Reproduced on a clean `e523bedc6` with two hand-made custom columns and no fork code involved. Setting a default value on the definition does not help — the panel reads the task's value, not the definition's.

I'll file it as a separate issue with the stack trace and a suggested patch. Mentioning it here because it decides how the effort work should land: my fork creates such columns, so it triggers this reliably, and a PR that adds them without the fix would make the problem visible for everyone.

On compensation — thank you, that's generous. Let's keep it separate so it doesn't hold up the work; I'll email you about it. I'd rather get the first PR in front of you.
